File: ostium-python-sdk/ostium_python_sdk/scscript/funding.py
import sys
from eth_abi import encode
from decimal import *
import math

PRECISION_2 = Decimal(1e2)
PRECISION_6 = Decimal(1e6)
PRECISION_18 = Decimal(1e18)
quantization_2 = Decimal('0.01')
quantization_3 = Decimal('0.001')
quantization_6 = Decimal('0.000001')
quantization_12 = Decimal('0.000000000001')
quantization_18 = Decimal('0.000000000000000001')

def getTargetFundingRate(normalizedOiDelta, hillInflectionPoint, maxFundingFeePerBlock, hillPosScale, hillNegScale):
    a = Decimal('1.84')
    n = 2
    K = Decimal('0.16')
    x = (a*normalizedOiDelta).quantize(quantization_6, rounding=ROUND_DOWN)
    x2 = (x**n).quantize(quantization_18, rounding=ROUND_DOWN)
    hill = (x2/(K + x2)).quantize(quantization_18, rounding=ROUND_DOWN)
    
    targetFr = (hillPosScale * hill).quantize(quantization_18, rounding=ROUND_DOWN) + hillInflectionPoint if normalizedOiDelta >= 0 \
    else (-hillNegScale * hill).quantize(quantization_18, rounding=ROUND_DOWN) + hillInflectionPoint
    
    if targetFr > Decimal('1'):
        targetFr = Decimal('1')
    elif targetFr < Decimal('-1'):
        targetFr = Decimal('-1')
    
    return (maxFundingFeePerBlock * targetFr).quantize(quantization_18, rounding=ROUND_DOWN)


def getPendingAccFundingFees(
        blockNumber: Decimal,
        lastUpdateBlock: Decimal,
        valueLong: Decimal, 
        valueShort: Decimal, 
        openInterestUsdcLong: Decimal, 
        openInterestUsdcShort: Decimal,
        OiCap: Decimal,        
        maxFundingFeePerBlock: Decimal,        
        lastFundingRate: Decimal,      
        hillInflectionPoint: Decimal,
        hillPosScale: Decimal,
        hillNegScale: Decimal,
        springFactor: Decimal,
        sFactorUpScale: Decimal,
        sFactorDownScaleP: Decimal,
    ):

    numBlocks = blockNumber - lastUpdateBlock
    openInterestMax = max(openInterestUsdcLong, openInterestUsdcShort)
    normalizedOiDelta = ((openInterestUsdcLong - openInterestUsdcShort).quantize(quantization_6, rounding=ROUND_DOWN) / max(OiCap, openInterestMax).quantize(quantization_6, rounding=ROUND_DOWN)).quantize(quantization_6, rounding=ROUND_DOWN)

    targetFr = getTargetFundingRate(normalizedOiDelta, hillInflectionPoint, maxFundingFeePerBlock, hillPosScale, hillNegScale)

    # New regime selection based on funding rates:
    if lastFundingRate * targetFr >= 0:  # Same sign
        if abs(targetFr) > abs(lastFundingRate):
            sFactor = springFactor
        else:
            sFactor = sFactorDownScaleP * springFactor / 100
    else:
        sFactor = sFactorUpScale * springFactor / 100

    expComp = exponentialApproximation(-sFactor * numBlocks)
    accFundingRate = (targetFr * numBlocks).quantize(quantization_18, rounding=ROUND_DOWN) + ((Decimal(1) - expComp) * (lastFundingRate - targetFr) / sFactor).quantize(quantization_18, rounding=ROUND_DOWN)

    fr = targetFr + ((lastFundingRate - targetFr) * expComp).quantize(quantization_18, rounding=ROUND_DOWN)
    
    if accFundingRate > 0:
        if openInterestUsdcLong > 0:
            valueLong += accFundingRate
            valueShort -= (accFundingRate * openInterestUsdcLong / openInterestUsdcShort).quantize(quantization_18, rounding=ROUND_DOWN) if openInterestUsdcShort > 0 else 0
    else:
        if openInterestUsdcShort > 0:
            valueShort -= accFundingRate
            valueLong += (accFundingRate * openInterestUsdcShort / openInterestUsdcLong).quantize(quantization_18, rounding=ROUND_DOWN) if openInterestUsdcLong > 0 else 0

    return (valueLong.quantize(quantization_18, rounding=ROUND_DOWN), valueShort.quantize(quantization_18, rounding=ROUND_DOWN), fr.quantize(quantization_18, rounding=ROUND_DOWN), targetFr.quantize(quantization_18, rounding=ROUND_DOWN))


def exponentialApproximation(value):
  if abs(value) < Decimal('0.7932312589092019'):
    numeratorTmp = value + Decimal('3');
    numerator = (numeratorTmp * numeratorTmp).quantize(quantization_18, rounding=ROUND_DOWN) + Decimal('3');
    denominatorTmp = value - Decimal('3');
    denominator = (denominatorTmp * denominatorTmp).quantize(quantization_18, rounding=ROUND_DOWN) + Decimal('3');

    return (numerator / denominator).quantize(quantization_18, rounding=ROUND_DOWN);
  elif abs(value) <= Decimal('6.906'):
    k = [Decimal('1.648721'), Decimal('1.284025'), Decimal('1.133148'), Decimal('1.064494'), Decimal('1.031743'), Decimal('1.015748'), Decimal('1.007843'), Decimal('1.003915'), Decimal('1.001955'), Decimal('1.000977')]
    product = Decimal('1.0')

    integer_part = math.floor(abs(value))
    decimal_part = abs(value) - integer_part
    
    for i in range(len(k)):
        decimal_part *= 2
        if decimal_part >= 1:
            product *= k[i]
            decimal_part -= 1
            product = product.quantize(quantization_6, rounding=ROUND_DOWN)
        if decimal_part == 0:
            break
    product = product.quantize(quantization_3, rounding=ROUND_DOWN) * Decimal(2)**(integer_part)

    return (Decimal(1) / product.quantize(quantization_18, rounding=ROUND_DOWN)).quantize(quantization_3, rounding=ROUND_DOWN)
  else:
    return Decimal(0)

if __name__ == "__main__":
    getcontext().prec = 128
    getcontext().rounding = ROUND_DOWN
    getcontext().clear_flags()

    if sys.argv[1] == "targetFundingRate":        
        normalizedOiDelta = Decimal(sys.argv[2]) / PRECISION_6
        hillInflectionPoint = Decimal(sys.argv[3]) / PRECISION_18
        maxFundingFeePerBlock = Decimal(sys.argv[4]) / PRECISION_18
        hillPosScale = Decimal(sys.argv[5]) / PRECISION_2
        hillNegScale = Decimal(sys.argv[6]) / PRECISION_2

        result = getTargetFundingRate(
            normalizedOiDelta,
            hillInflectionPoint,
            maxFundingFeePerBlock,
            hillPosScale,
            hillNegScale
        )
        encodedResult = "0x" + \
            encode(["int"], [int(result*PRECISION_18)]).hex()
        print(encodedResult)
    elif sys.argv[1] == "accFundingRate":
        blockNumber = Decimal(sys.argv[2])
        lastUpdateBlock = Decimal(sys.argv[3])
        valueLong = Decimal(sys.argv[4]) / PRECISION_18
        valueShort = Decimal(sys.argv[5]) / PRECISION_18
        openInterestUsdcLong = Decimal(sys.argv[6]) / PRECISION_6
        openInterestUsdcShort = Decimal(sys.argv[7]) / PRECISION_6
        oiCap = Decimal(sys.argv[8]) / PRECISION_6
        maxFundingFeePerBlock = Decimal(sys.argv[9]) / PRECISION_18
        lastFundingRate = Decimal(sys.argv[10]) / PRECISION_18
        hillInflectionPoint = Decimal(sys.argv[11]) / PRECISION_18
        hillPosScale = Decimal(sys.argv[12]) / PRECISION_2
        hillNegScale = Decimal(sys.argv[13]) / PRECISION_2
        springFactor = Decimal(sys.argv[14]) / PRECISION_18
        sFactorUpScale = Decimal(sys.argv[15]) / PRECISION_2
        sFactorDownScaleP = Decimal(sys.argv[16]) / PRECISION_2

        result = getPendingAccFundingFees(blockNumber, lastUpdateBlock, valueLong, valueShort, openInterestUsdcLong,
                                        openInterestUsdcShort, oiCap, maxFundingFeePerBlock, lastFundingRate,
                                        hillInflectionPoint, hillPosScale, hillNegScale, springFactor, sFactorUpScale, sFactorDownScaleP)
        encodedResult = "0x" + \
            encode(["int", "int", "int64"], [int(result[0]*PRECISION_18),
                int(result[1]*PRECISION_18), int(result[2]*PRECISION_18)]).hex()
        print(encodedResult)
    else:
        print("Invalid Operation")
        sys.exit(1)      

File: src/formulae/funding.rs
use rust_decimal::prelude::ToPrimitive;
use rust_decimal::{Decimal, MathematicalOps, RoundingStrategy};
use rust_decimal_macros::dec;

const RD: RoundingStrategy = RoundingStrategy::ToZero;

pub fn get_target_funding_rate(
    normalized_oi_delta: Decimal,
    hill_inflection_point: Decimal,
    max_funding_fee_per_block: Decimal,
    hill_pos_scale: Decimal,
    hill_neg_scale: Decimal,
) -> Decimal {
    let a = dec!(1.84);
    let n = 2u64;
    let k = dec!(0.16);
    let x = (a * normalized_oi_delta).round_dp_with_strategy(6, RD);
    let x2 = x.powi(n as i64).round_dp_with_strategy(18, RD);
    let hill = (x2 / (k + x2)).round_dp_with_strategy(18, RD);

    let target_fr = if normalized_oi_delta >= Decimal::ZERO {
        (hill_pos_scale * hill).round_dp_with_strategy(18, RD) + hill_inflection_point
    } else {
        (-hill_neg_scale * hill).round_dp_with_strategy(18, RD) + hill_inflection_point
    };
    let target_fr = target_fr.clamp(-Decimal::ONE, Decimal::ONE);

    (max_funding_fee_per_block * target_fr).round_dp_with_strategy(18, RD)
}

pub fn get_pending_acc_funding_fees(
    block_number: Decimal,
    last_update_block: Decimal,
    mut value_long: Decimal,
    mut value_short: Decimal,
    open_interest_usdc_long: Decimal,
    open_interest_usdc_short: Decimal,
    oi_cap: Decimal,
    max_funding_fee_per_block: Decimal,
    last_funding_rate: Decimal,
    hill_inflection_point: Decimal,
    hill_pos_scale: Decimal,
    hill_neg_scale: Decimal,
    spring_factor: Decimal,
    s_factor_upscale: Decimal,
    s_factor_down_scale_p: Decimal,
) -> (Decimal, Decimal, Decimal, Decimal) {
    let num_blocks = block_number - last_update_block;
    let open_interest_max = open_interest_usdc_long.max(open_interest_usdc_short);

    let denominator = oi_cap.max(open_interest_max).round_dp_with_strategy(6, RD);
    let norm_oi_delta = ((open_interest_usdc_long - open_interest_usdc_short)
        .round_dp_with_strategy(6, RD)
        / denominator)
        .round_dp_with_strategy(6, RD);

    let target_fr = get_target_funding_rate(
        norm_oi_delta,
        hill_inflection_point,
        max_funding_fee_per_block,
        hill_pos_scale,
        hill_neg_scale,
    );

    let hundred = Decimal::from(100);
    let s_factor = if (last_funding_rate * target_fr) >= Decimal::ZERO {
        if target_fr.abs() > last_funding_rate.abs() {
            spring_factor
        } else {
            (s_factor_down_scale_p * spring_factor) / hundred
        }
    } else {
        (s_factor_upscale * spring_factor) / hundred
    };

    let exp_input = -(s_factor * num_blocks);
    let exp_comp = exp_approx(exp_input);

    let term1 = (target_fr * num_blocks).round_dp_with_strategy(18, RD);
    let term2 = ((Decimal::ONE - exp_comp) * (last_funding_rate - target_fr) / s_factor)
        .round_dp_with_strategy(18, RD);
    let acc_funding_rate = term1 + term2;

    let fr = target_fr + ((last_funding_rate - target_fr) * exp_comp).round_dp_with_strategy(18, RD);

    if acc_funding_rate > Decimal::ZERO {
        if open_interest_usdc_long > Decimal::ZERO {
            value_long += acc_funding_rate;
            if open_interest_usdc_short > Decimal::ZERO {
                let short_adj = (acc_funding_rate * open_interest_usdc_long
                    / open_interest_usdc_short)
                    .round_dp_with_strategy(18, RD);
                value_short -= short_adj;
            }
        }
    } else if acc_funding_rate < Decimal::ZERO {
        if open_interest_usdc_short > Decimal::ZERO {
            value_short -= acc_funding_rate; // subtracting negative adds to value
            if open_interest_usdc_long > Decimal::ZERO {
                let long_adj = (acc_funding_rate * open_interest_usdc_short
                    / open_interest_usdc_long)
                    .round_dp_with_strategy(18, RD);
                value_long += long_adj;
            }
        }
    }

    (
        value_long.round_dp_with_strategy(18, RD),
        value_short.round_dp_with_strategy(18, RD),
        fr.round_dp_with_strategy(18, RD),
        target_fr.round_dp_with_strategy(18, RD),
    )
}

/// Padé-style + lookup-table exponential approximation matching the on-chain implementation.
pub fn exp_approx(val: Decimal) -> Decimal {
    if val.abs() < dec!(0.7932312589092019) {
        let three = Decimal::from(3);
        let n_tmp = val + three;
        let numerator = (n_tmp * n_tmp).round_dp_with_strategy(18, RD) + three;
        let d_tmp = val - three;
        let denominator = (d_tmp * d_tmp).round_dp_with_strategy(18, RD) + three;
        return (numerator / denominator).round_dp_with_strategy(18, RD);
    } else if val.abs() <= dec!(6.906) {
        let k_values = [
            dec!(1.648721),
            dec!(1.284025),
            dec!(1.133148),
            dec!(1.064494),
            dec!(1.031743),
            dec!(1.015748),
            dec!(1.007843),
            dec!(1.003915),
            dec!(1.001955),
            dec!(1.000977),
        ];

        let mut product = Decimal::ONE;
        let abs_val = val.abs();
        let integer_part = abs_val.round_dp_with_strategy(0, RD);
        let mut decimal_part = abs_val - integer_part;
        let two = Decimal::from(2);

        for k in k_values.iter() {
            decimal_part *= two;
            if decimal_part >= Decimal::ONE {
                product = (product * *k).round_dp_with_strategy(6, RD);
                decimal_part -= Decimal::ONE;
            }
            if decimal_part == Decimal::ZERO {
                break;
            }
        }

        let pow_val = integer_part.to_u64().unwrap_or(0);
        let multiplier = two.powi(pow_val as i64);

        product = product.round_dp_with_strategy(3, RD) * multiplier;

        let final_denominator = product.round_dp_with_strategy(18, RD);
        return (Decimal::ONE / final_denominator).round_dp_with_strategy(3, RD);
    } else {
        Decimal::ZERO
    }
}
